### PR TITLE
Complete Post page backend. Generalized firestore read / listen functions.

### PR DESCRIPTION
**Pull request by saraholiviawu**
**Complete Post page backend**: 
  -  **Write data to Firestore**: Store blob key, caption, user name, business name, timestamp in a 'posts' collection within Firestore.
  - **Filtered Firestore documents by city & state and ordered by most recent to oldest**: When the posts page loads, a query is sent to the Firebase Firestore collection 'posts' to get documents that match this city and state. The posts are then ordered in descending timestamp. 
  - **Read Firestore data to construct a post**: The user id of the post's owner is read from each post that matches the queries made and used to retrieve the business name from the 'businesses' collection as well as the user name from 'users' collection. Because Firestore queries return promises, it is necessary to nest these queries to ensure that all data has been retrieved from a single collection before moving to the next. 
  - **Dynamically construct each post**: A post element is populated with data and added to the HTML page dynamically through JQuery. 
  - **Save caption to local storage if post flow is interrupted**: Save caption to local storage if user clicks away from post or uploads a new photo. Local storage item is deleted once the user uploads the post or if there is no blob key present in the URL (ex: occurs if user navigates to another page and comes back after writing a caption without posting). 

**Generalized firestore read / listen function**: Previously, to read, filter, or listen to Firestore, each function needed to call Firebase Firestore directly and customize read or listen method / filters / ordering based on use case. Currently:
  - **For the function getOrSnapshotDocsByQuery( )**: 
    - makes a get() request OR listens for live updates via the Snapshot method
    - filters for documents based on a document's fields, which can be specified as a list in the parameter
    - [optional parameter] Sorts query by a specified order; otherwise sorts in ascending order
    - takes in a lambda function that handles what should be done after the desired documents are retrieved
  - **For the function getDocByDocId( )**:
    - Get a doc by specifying a doc id, NOT filter values or ordering

**Demo**:
1. User adds a post and can see it appear on the page immediately.
See full video:https://drive.google.com/file/d/1oEGCshsxK6PlGn6eY28LJRAI6mpQNRn4/view
Backend  Add new post to posts html

2. User can see a draft caption even after clicking away or upload a new image. User will not see the saved caption after sharing a post. 
See full video: https://drive.google.com/file/d/14Y--g6J8BpUTeAgNqqpzSJD3ukLULRcV/view

### CODE REVIEW

File: ripple/src/main/webapp/auth.js [hunk 1]
```diff
@@ -176,14 +168,14 @@ function signOutUser(redirectPage="index.html") {
 Passes blobKey to getBlobKey() */
 function getAcctInfo(uid) {
   console.log("Success: getAcctInfo() recognizes uid: " + uid);
-  var name, email, isBusinessOwner, address, blobKey;
+  var userName, email, isBusinessOwner, address, blobKey;
   db.collection("users")
       .where("uid", "==", uid)
       .get()
       .then(function(querySnapshot) {
```
Comment (peter14f) [inline ripple/src/main/webapp/auth.js:175]:
```
      .then((querySnapshot) => {
```

File: ripple/src/main/webapp/posts.js [hunk 2]
```diff
@@ -23,50 +32,105 @@ function postsOnload() {
...
+  var uid = localStorage.getItem("uid");
+  console.log("uid in newPost(): " + uid)
+  var blobKey = readBlobKeyFromURl();
+  var caption = document.getElementById("modal-caption").value;
+  var varArray = [uid, blobKey, caption];
+  var whereFieldArray = ["uid"];
+  var whereValueArray = [uid];
+  // Get city and state from businesses
+  var lambda = function(querySnapshot) {
+    querySnapshot.forEach(function(doc) {
```
Comment (peter14f) [inline ripple/src/main/webapp/posts.js:60]:
```
    querySnapshot.forEach((doc) => {
```

File: ripple/src/main/webapp/posts.js [hunk 3]
```diff
@@ -23,50 +32,105 @@ function postsOnload() {
...
+function addPostToFirestore(collection, uid, blobKey, caption, city, state) {
+  console.log(collection + " " + uid);
+  db.collection('posts').add({
+    uid: uid,
+    postBlobKey: blobKey,
+    caption: caption,
+    city: city,
+    state: state,
+    timestamp: firebase.firestore.FieldValue.serverTimestamp(),
+  }).then(function() {
```
Comment (peter14f) [inline ripple/src/main/webapp/posts.js:81]:
```
  }).then(() => {
```

File: ripple/src/main/webapp/posts.js [hunk 4]
```diff
@@ -23,50 +32,105 @@ function postsOnload() {
...
+    window.location = 'postcommunity.html';
+    console.log("Document successfully written!");
+  }).catch(function(error) {
```
Comment (peter14f) [inline ripple/src/main/webapp/posts.js:84]:
```
  }).catch((error) => {
```

File: ripple/src/main/webapp/posts.js [hunk 5]
```diff
@@ -23,50 +32,105 @@ function postsOnload() {
...
+      console.log("postBlobKey: " + postBlobKey);
+      var caption = doc.data().caption;
+      console.log("caption: " + caption);
+      // DELETE FOR FINAL PROD
+      var postOrder = doc.data().postOrder;
+      console.log("postOrder: " + postOrder);
+      var whereFieldArray = ["uid"];
+      var whereValueArray = [uid];
+      // Get businessName from "businesses" collection using uid. Assume one business per business owner max.
+      var secondLambda = function(secondQuerySnapshot) {
```
Comment (peter14f) [inline ripple/src/main/webapp/posts.js:109]:
```
      var secondLambda =(secondQuerySnapshot) => {
```

File: ripple/src/main/webapp/posts.js [hunk 6]
```diff
@@ -23,50 +32,105 @@ function postsOnload() {
...
+        secondQuerySnapshot.forEach(function(doc) {
```
Comment (peter14f) [inline ripple/src/main/webapp/posts.js:110]:
```
        secondQuerySnapshot.forEach((doc) => {
```

File: ripple/src/main/webapp/posts.js [hunk 7]
```diff
@@ -23,50 +32,105 @@ function postsOnload() {
...
+          var businessName = doc.data().businessName[1];
+          console.log("businessName: " + businessName);
+          // Get userName from "users" collection using uid
+          var thirdLambda = function(thirdQuerySnapshot) {
```
Comment (peter14f) [inline ripple/src/main/webapp/posts.js:114]:
```
          var thirdLambda =(thirdQuerySnapshot) => {
```

File: ripple/src/main/webapp/posts.js [hunk 8]
```diff
@@ -23,50 +32,105 @@ function postsOnload() {
...
+            thirdQuerySnapshot.forEach(function(doc) {
```
Comment (peter14f) [inline ripple/src/main/webapp/posts.js:115]:
```
            thirdQuerySnapshot.forEach((doc) => {
```

File: ripple/src/main/webapp/script.js [hunk 9]
```diff
@@ -50,6 +51,48 @@ function getParameterByName(name) {
   return decodeURIComponent(results[2].replace(/\+/g, ' '));
 }
 
+/* Retrieve a doc by docId */
+function getDocByDocId(collection, docId, lambda, varArray=false) {
+  db.collection(collection).doc(docId).get()
+  .then(lambda)
+  .catch(function(error) {
```
Comment (peter14f) [inline ripple/src/main/webapp/script.js:58]:
```
  .catch((error) => {
```

File: ripple/src/main/webapp/script.js [hunk 10]
```diff
@@ -50,6 +51,48 @@ function getParameterByName(name) {
...
+      ref = ref.orderBy(orderByField);
+    }
+  }
+  // Get documents based on method get or snapshot
+  if (method == "get") {
+    ref.get()
+    .then(lambda);
+  } else { // method == "snapshot"
+    ref.onSnapshot(lambda)
+    .catch(function(error) {
```
Comment (peter14f) [inline ripple/src/main/webapp/script.js:90]:
```
    .catch((error) => {
```

File: ripple/src/main/webapp/auth.js (hunk 1 above)
Comment (saraholiviawu) [reply ripple/src/main/webapp/auth.js:175]: Fixed. This function is used for Account Settings page and will be modified in the future / migrated to an accountInfo.js file.

File: ripple/src/main/webapp/script.js (hunk 10 above)
Comment (saraholiviawu) [reply ripple/src/main/webapp/script.js:90]: Fixed

File: ripple/src/main/webapp/script.js (hunk 9 above)
Comment (saraholiviawu) [reply ripple/src/main/webapp/script.js:58]: Fixed

File: ripple/src/main/webapp/posts.js (hunk 8 above)
Comment (saraholiviawu) [reply ripple/src/main/webapp/posts.js:115]: Fixed

File: ripple/src/main/webapp/posts.js (hunk 7 above)
Comment (saraholiviawu) [reply ripple/src/main/webapp/posts.js:114]: Fixed

File: ripple/src/main/webapp/posts.js (hunk 6 above)
Comment (saraholiviawu) [reply ripple/src/main/webapp/posts.js:110]: Fixed

File: ripple/src/main/webapp/posts.js (hunk 5 above)
Comment (saraholiviawu) [reply ripple/src/main/webapp/posts.js:109]: Fixed

File: ripple/src/main/webapp/posts.js (hunk 4 above)
Comment (saraholiviawu) [reply ripple/src/main/webapp/posts.js:84]: Fixed

File: ripple/src/main/webapp/posts.js (hunk 3 above)
Comment (saraholiviawu) [reply ripple/src/main/webapp/posts.js:81]: Fixed

File: ripple/src/main/webapp/posts.js (hunk 2 above)
Comment (saraholiviawu) [reply ripple/src/main/webapp/posts.js:60]: Fixed

**Pull request merged**